File: src/com/brianc/css/Specificity.java
package com.brianc.css;

import java.util.Comparator;

public class Specificity implements Comparable<Specificity> {
	int a;
	int b;
	int c;
	
	private static Comparator<Specificity> cmp;
	static {
		// HACK to placate Java's type inference
		cmp = Comparator.comparingInt(s -> s.a);
		cmp = cmp.thenComparingInt(s -> s.b).thenComparingInt(s -> s.c);
	}

	public Specificity(int a, int b, int c) {
		this.a = a;
		this.b = b;
		this.c = c;
	}

	@Override
	public int compareTo(Specificity other) {
		return cmp.compare(this, other);
	}
}

File: src/com/brianc/css/Specificity.kt
package com.brianc.css

import kotlin.comparisons.compareValuesBy

data class Specificity(val a: Int, val b: Int, val c: Int) : Comparable<Specificity> {
    override fun compareTo(other: Specificity) = compareValuesBy(this, other, { it.a }, { it.b }, { it.c })
}
